### ISSUE TITLE
[Glimmer] helper no longer receiving `env`

### ISSUE THREAD

**Issue by rlivsey** [emberjs/ember.js#11080]
In [ember-can](https://github.com/minutebase/ember-can) we've got a helper which expects to receive `env` but since Glimmer merged it no longer is.

The is [defined as so](https://github.com/minutebase/ember-can/blob/master/addon/helpers/can.js#L30-35):
```js
export default function(properties, hash, options, env) {
  var abilityName  = properties[0];
  var resource     = properties[1];
  var view         = env.data.view;
  var container    = view.container;
```

and is then [exported as a plain function](https://github.com/minutebase/ember-can/blob/master/app/helpers/can.js):
```js
import canHelper from 'ember-can/helpers/can';
export default canHelper;
```

This may well be an issue in the addon and shouldn't have received `env` previously, but before I fix it in the addon itself I thought it worth checking if this is a regression or if there's a different way we should be doing this.

**Comment by rlivsey** [comment]
Just realised I missed a part out, we're manually registering the helper (because it's not hyphenated) [in an initializer](https://github.com/minutebase/ember-can/blob/master/app/initializers/setup-ember-can-helpers.js):

```
import Ember from 'ember';

import helper from '../helpers/can';

export default {
  name: 'setup-ember-can-helpers',
  initialize: function() {
    Ember.HTMLBars._registerHelper('can', helper);
  }
};
```

This is using the private `_regsterHelper`, so I imagine there is no regression here as it wasn't public API.

**Comment by rwjblue** [comment]
> This is using the private _regsterHelper, so I imagine there is no regression here as it wasn't public API.

Yes, it is unfortunate but true.  I am sorry :disappointed:

**Comment by rlivsey** [comment]
@rwjblue is there a different way we should be getting `env` or the `container` here?

**Comment by wycats** [comment]
@rlivsey helpers aren't going to have access to the ambient environment anymore (making sure they reliably update is extremely finicky), but I have some thoughts on how to make a helper like `ember-can` possible.

There are a few different options, and we can pick a good one once we see a bigger set of the private-API "regressions" that Glimmer caused.

**Comment by rlivsey** [comment]
@wycats for now I've updated `ember-can` to use an `if-can/unless-can` component which covers 99% of use-cases for now.

It means you can no longer compose with liquid-fire (`liquid-if (can "foo")`) or use it to easily change classes (`class="{{if (can "bar") "some-class"}}"`) but those are pretty rare anyway.

[ember-intl](https://github.com/yahoo/ember-intl) is another addon in which I've run into this issue, they rely on `env` in the helper to get `container` to lookup the locales & translations service.

**Comment by knownasilya** [comment]
Both sound like legitimate use cases for accessing env/container.

**Comment by amk221** [comment]
+1 I also need to lookup localisation data

**Comment by rwjblue** [comment]
Just wanted to mention ember-cli-i18n is also in the mix here. See https://github.com/dockyard/ember-cli-i18n/issues/106.

**Comment by amk221** [comment]
Could somebody mention the potential ideas for accessing the container from a helper? I'm pretty much stuck until I can figure this out.. and I don't want to set off on the wrong foot. Thanks!

**Comment by rlivsey** [comment]
@amk221 it's essentially 'impossible' right now (where by impossible I mean you can technically do it through a few nefarious means, but they're not a good idea) so AFAICT it's a case of the core team gathering use-cases and seeing what the best public API is to suit these needs.

**Comment by cibernox** [comment]
Is there any way to track or help with the issue of exposing the container? 

There is several *extremely* popular addons that cannot be updated to work with glimmer without an idiomatic way of using the container.

A good example is `ember-cli-i18n`, that probably is preventing a few thousands of devs with multilingual apps of trying glimmer.

**Comment by mixonic** [comment]
@cibernox I'm very aware of the pain. I'll make it my mission today to put together a draft of a proposal for `Expression` we've kicked around in chat. I'm 100% sure it is the solution for all ills, but I am optimistic.

**Comment by mixonic** [comment]
@cibernox and all, I've put some API suggestions that would help with these use-cases into an RFC: https://github.com/emberjs/rfcs/pull/52

**Comment by elwayman02** [comment]
+1, this issue is blocking my ability to upgrade to glimmer due to it breaking ember-intl. :cry:

**Comment by DingoEatingFuzz** [comment]
I also just got bit by this due to the modal we have. I do believe the Stateful Helpers RFC (https://github.com/emberjs/rfcs/pull/53) will solve this issue.

**Comment by rwjblue** [comment]
Yes, that RFC is intending to solve this issue.

**Comment by pierrickrouxel** [comment]
I use `options.data.view` to do a createChildView() into an helper. This allows me to use the bootstrap popover with the current context.

You can see sources here: [https://github.com/pierrickrouxel/ember-strap/blob/master/addon/helpers/popover.js](https://github.com/pierrickrouxel/ember-strap/blob/master/addon/helpers/popover.js)

Do you have an idea to how can I port it to ember 1.13.0. I don't find any way to continue to use my component.

**Comment by rwjblue** [comment]
Closing as the initial issue was addressed in 1.13.0 via [emberjs/rfcs#53](https://github.com/emberjs/rfcs/blob/master/text/0053-helpers.md) and [emberjs/rfcs#58](https://github.com/emberjs/rfcs/blob/master/text/0058-helper-listing.md).

@pierrickrouxel - That doesn't seem directly related, and it seems more like something suited for #needhelp in the [community slack](https://ember-community-slackin.herokuapp.com/) or StackOverflow.
